sklearn/datasets/openml.py

```diff
@@ -0,0 +1,470 @@
+
+    Parameters
+    ----------
+    name : str
+        name of the dataset
+
+    version : int or str
+        If version is an integer, the exact name/version will be obtained from
+        OpenML. If version is a string (value: "active") it will take the first
+        version from OpenML that is annotated as active. Any other string
+        values except "active" are treated as integer.
+
+    Returns
+    -------
+    first_dataset : json
+        json representation of the first dataset object that adhired to the
+        search criteria
+
+    """
+    if version == "active":
+        # situation in which we return the oldest active version
+        url = _SEARCH_NAME.format(name) + "/status/active/"
+        error_msg = "No active dataset {} found.".format(name)
+        json_data = _get_json_content_from_openml_api(url, error_msg, True)
+        return json_data['data']['dataset'][0]
+
+    # an integer version has been provided
+    url = (_SEARCH_NAME + "/data_version/{}").format(name, version)
+    json_data = _get_json_content_from_openml_api(url, None, False)
+    if json_data is None:
+        # we can do this in 1 function call if OpenML does not require the
+        # specification of the dataset status (i.e., return datasets with a
+        # given name / version regardless of active, deactivated, etc. )
+        # TODO: feature request OpenML.
+        url += "/status/deactivated"
+        error_msg = "Dataset {} with version {} not found.".format(name,
+                                                                   version)
+        json_data = _get_json_content_from_openml_api(url, error_msg, True)
+
+    return json_data['data']['dataset'][0]
+
+
+def _get_data_description_by_id(data_id):
+    # OpenML API function: https://www.openml.org/api_docs#!/data/get_data_id
+    url = _DATA_INFO.format(data_id)
+    error_message = "Dataset with data_id {} not found.".format(data_id)
+    json_data = _get_json_content_from_openml_api(url, error_message, True)
+    return json_data['data_set_description']
+
+
+def _get_data_features(data_id):
+    # OpenML function:
+    # https://www.openml.org/api_docs#!/data/get_data_features_id
+    url = _DATA_FEATURES.format(data_id)
+    error_message = "Dataset with data_id {} not found.".format(data_id)
+    json_data = _get_json_content_from_openml_api(url, error_message, True)
+    return json_data['data_features']['feature']
+
+
+def _download_data_arff(file_id, sparse):
+    # Accesses an ARFF file on the OpenML server. Documentation:
+    # https://www.openml.org/api_data_docs#!/data/get_download_id
+    url = _DATA_FILE.format(file_id)
+    response = urlopen(url)
+    kwargs = {}
+    if sparse is True:
+        kwargs['return_type'] = arff.COO
+    else:
+        kwargs['return_type'] = arff.DENSE
+
+    if PY2:
+        # Because Python2.7 numpy can't handle unicode
+        arff_file = arff.loads(response.read(), **kwargs)
```

Comment:
Is `arff.load` not more efficient than `arff.loads`?

Reply:
I would've thought `response.read()` here could sometimes overwhelm memory.

Reply:
It probably is, I will adapt the code